libraries/AC_SafeRTL/AC_SafeRTL.cpp

```diff
@@ -0,0 +1,573 @@
+                float dist = point_line_dist(_path[i], _path[start_index], _path[end_index]);
+                if (dist > max_dist) {
+                    index = i;
+                    max_dist = dist;
+                }
+            }
+        }
+
+        if (max_dist > SAFERTL_SIMPLIFICATION_EPSILON) {
+            // if the to-do list is full, give up on simplifying. This should never happen.
+            if (_simplification_stack_last_index > _current_path_len * SAFERTL_SIMPLIFICATION_STACK_LEN_MULT) {
+                _simplification_complete = true;
+                return;
+            }
+            _simplification_stack[++_simplification_stack_last_index] = start_finish {start_index, index};
+            _simplification_stack[++_simplification_stack_last_index] = start_finish {index, end_index};
+        } else {
+            for (int16_t i = start_index + 1; i < end_index; i++) {
+                _simplification_bitmask[i] = false;
+            }
+        }
+    }
+    _simplification_complete = true;
+}
+
+/**
+*   This method runs for the allotted time, and detects loops in a path. All detected loops are added to _prunable_loops,
+*   this function does not alter the path in memory. It works by comparing the line segment between any two sequential points
+*   to the line segment between any other two sequential points. If they get close enough, anything between them could be pruned.
+*
+*   Note that this method might take a bit longer than LOOP_TIME. It only stops after it's already run longer.
+*/
+void AC_SafeRTL::detect_loops()
+{
+    // if SafeRTL is not active OR if this algorithm has already run to completion OR there's fewer than 3 points in the path
+    if (!_active || _pruning_complete || _last_index < 3) {
+        return;
+    }
+    uint32_t start_time = AP_HAL::micros();
+
+    while (_pruning_current_i < _last_index - 1) {
+        // if this method has run for long enough, exit
+        if (AP_HAL::micros() - start_time > SAFERTL_LOOP_TIME_US) {
+            return;
+        }
+
+        // this check prevents detection of a loop-within-a-loop
+        int16_t j = MAX(_pruning_current_i + 2, _pruning_min_j);
+        while (j < _last_index) {
+            dist_point dp = segment_segment_dist(_path[_pruning_current_i], _path[_pruning_current_i+1], _path[j], _path[j+1]);
+            if (dp.distance <= SAFERTL_PRUNING_DELTA) { // if there is a loop here
+                _pruning_min_j = j;
+                // if the buffer is full
+                if ( _prunable_loops_last_index >= _current_path_len * SAFERTL_LOOP_BUFFER_LEN_MULT - 1) {
+                    _pruning_complete = true; // pruning is effectively complete now, since there's no reason to continue looking for them.
+                    return;
+                }
+                // int promotion rules disallow using i+1 and j+1
+                _prunable_loops[++_prunable_loops_last_index] = {(++_pruning_current_i)--,(++j)--,dp.point};
```

Comment:
The line is supposed to look like this: 
`_prunable_loops[++_prunable_loops_last_index] = {_pruning_current_i + 1, j + 1, dp.point};`.

The problem is that apparently adding 1 to a `uint8_t` will promote the `uint8_t` to a `uint_32_t` before adding one. Casting back down gives annoying compiler errors.